Rewrite manifest, storagemanifest, bufmanifest packages into bufcas package

bufdev opened the pull request.

unmultimedio reviewed (commented):
Looking very nice in general, left a few comments and questions. For the cache I think @pkwarren eyes are best for it.

The only thing I'm missing is the tests we had for both digests, blobs, and manifests in the previous pkgs. Do you plan on migrating those too?

bufdev commented on private/bufpkg/bufcas/file_node.go:
```diff
@@ -0,0 +1,140 @@
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+)
+
+// FileNode is a path and associated digest.
+type FileNode interface {
```
This is intended to match the naming of the registry API - `Digest, Blob, FileNode` all have direct corollaries, and since we don't have the concept of `Manifest` in the registry API, I'd like to keep it like this.

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+	// The paths of the given FileNodes are guaranteed to be unique.
+	// The iteration order will be the sorted order of the paths.
+	FileNodes() []FileNode
+	// GetDigest gets the Digest for the given path.
+	//
+	// Returns nil if the path does not exist.
+	GetDigest(path string) Digest
```
`*For` isn't a pattern anywhere else in the codebase, `Get*` is, re: my desire to keep things consistent.

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+		return nil, errors.New("string for Manifest did not end with newline")
+	}
+	s = s[:len(s)-1]
+	for i, line := range strings.Split(s, "\n") {
+		fileNode, err := NewFileNodeForString(line)
+		if err != nil {
+			return nil, fmt.Errorf("invalid Manifest at line %d: %w", i, err)
+		}
+		fileNodes = append(fileNodes, fileNode)
+	}
```
Hmm...I might argue we don't actually want this validation. It's borderline, but when constructing a `Manifest`, might be worth accepting strings that are unsorted. In `NewBlobSet`, for example, we deduplicate, and I thought about verifying that the input is already deduplicated, but it seemed good to accept it.

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+func (m *manifest) String() string {
+	buffer := bytes.NewBuffer(nil)
```
`bytes.NewBuffer(nil)` is the standard in the `buf` codebase:

```
$ agg bytes.newbuffer | wc -l
     105
```

bufdev commented on private/bufpkg/bufmodule/bufmodule.go:
```diff
@@ -133,21 +133,14 @@ type Module interface {
-	BlobSet() *manifest.BlobSet
+	// FileSet returns the FileSet for the module.
+	//
+	// FileSet will be nil if the Module was not constructed with a FileSet.
```
Yes - this was the original motivation for all of this work :-) This comment will _definitely_ be deleted in the very near future, but this PR is just trying to get all the plumbing working with the existing setup.

pkwarren commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+	return BlobToProto(blob)
+}
+
+// BlobToManifest converts the given proto Blob representing the string representation of a Manifest into a Manifest.
```
```suggestion
// ProtoBlobToManifest converts the given proto Blob representing the string representation of a Manifest into a Manifest.
```

bufdev replied:
Fixed.

pkwarren commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+	for _, fileNode := range fileNodes {
+		if existingFileNode, ok := pathToFileNode[fileNode.Path()]; ok {
+			if !DigestEqual(existingFileNode.Digest(), fileNode.Digest()) {
+				return nil, fmt.Errorf("path %q had different Digests when constructing FileNode", fileNode.Path())
```
```suggestion
				return nil, fmt.Errorf("path %q had different Digests when constructing Manifest", fileNode.Path())
```

bufdev replied:
Done.

pkwarren commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+		_, _ = buffer.WriteString(fileNode.String())
+		_, _ = buffer.WriteRune('\n')
+	}
+	return buffer.String()
```
nit: Since we're going to use this a lot, we might switch to `strings.Builder` instead. It is documented to do less copying that bytes.Buffer for building strings.

Ref: https://pkg.go.dev/bytes#Buffer.String

pkwarren commented on private/pkg/storage/bucket.go:
```diff
@@ -82,6 +85,11 @@ func PutWithChunkSize(sizeInBytes int64) PutOption {
+// TODO: when do we NOT want this to be true? I believe a ton of the Put calls
+// in the buf codebase assume this is true, and were never refactored to take
+// into account that this might not be true. This should likely be removed,
+// and if anything, replaced with a "PutWithoutAtomic".
```
This is originally only needed for the CAS code that uses the OS - it helps to guarantee files are written atomically so we never read a partially written file.

The only negative would be if we forget a call to Close() (or there is a kill -9'd process) - in this cases we'd leave behind a temporary file on disk.

When rolling this out, it felt safer to keep it contained to one place to start. I'm ok with it being made the default.

bufdev replied:
Talked about this offline, removed TODOs for now.

pkwarren reviewed (commented):
This looks awesome - such a nice refactor. I still need to review the test failure in more detail to see if I can tell why it is failing.

pkwarren commented on private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go:
```diff
@@ -93,36 +97,28 @@ func (c *casModuleCacher) PutModule(
 			return fmt.Errorf("invalid module pin digest %q: %w", modulePinDigestEncoded, err)
 		}
-		if !manifestDigest.Equal(*modulePinDigest) {
+		if bufcas.DigestEqual(manifestDigest, modulePinDigest) {
```
```suggestion
		if !bufcas.DigestEqual(manifestDigest, modulePinDigest) {
```

This is the reason for the failing test.

bufdev replied:
Dang, fixed.

bufdev commented on private/bufpkg/bufcas/blob.go:
```diff
@@ -0,0 +1,155 @@
+//
+// The reader is read until io.EOF.
+// Validation is performed to ensure that the Digest matches the computed Digest of the content.
+func NewBlobForContentWithKnownDigest(knownDigest Digest, reader io.Reader) (Blob, error) {
```
Updated to this. I also updated `NewDigestForContent` to take `DigestWithDigestType`, and adjusted all the `ForString` functions to be `Parse` instead, matching the rest of the codebase (and also distinguishing that these functions do not take constructor options).

bufdev commented on private/bufpkg/bufcas/blob.go:
```diff
@@ -0,0 +1,155 @@
+// TODO: In the former version of this package, we compared content values as well.
+// We should be able to remove this, and this is commented out for now. Technically we do not
+// need to compare the contents, as we know that the Digest is a valid Digest for the
+// given content via valiation we did at construction time.
```
Fixed.

bufdev commented on private/bufpkg/bufcas/blob.go:
```diff
@@ -0,0 +1,155 @@
+// TODO: In the former version of this package, we compared content values as well.
+// We should be able to remove this, and this is commented out for now. Technically we do not
+// need to compare the contents, as we know that the Digest is a valid Digest for the
+// given content via valiation we did at construction time.
```
Moved out of a TODO and to an inline comment in the function, for our future reference.

bufdev commented on private/bufpkg/bufcas/blob_set.go:
```diff
@@ -0,0 +1,121 @@
+func newBlobSet(blobs []Blob) *blobSet {
+	digestStringToBlob := make(map[string]Blob, len(blobs))
+	sortedDigestStrings := make([]string, 0, len(blobs))
+	for _, blob := range blobs {
+		digestString := blob.Digest().String()
```
Concurring with @saquibmian: it should definitely be an error, and while I'm overall team-anti-`panic`, this codebase generally does not do such nil checking, as we generally assume as a principle that you need to have non-nil values unless documented otherwise. nil error checking everything in this codebase in such a case would be overly burdensome, and given that its a fundamental code error, we choose to rely on panic here.

bufdev commented on private/bufpkg/bufcas/digest.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+func newDigest(digestType DigestType, value []byte) (*digest, error) {
+	switch digestType {
+	case DigestTypeShake256:
```
I would argue otherwise in this case: generally I think it's good to set up `switches` for enums, as its a clear indication to future developers that this is a forked codepath based on the value of the enum. An `if` statement isn't as clear, and in my opinion having an `if` statement in these cases leads to future bugs when new enum values are added (the developer does not realize the `if` statement needs adjusting).

bufdev commented on private/bufpkg/bufcas/digest.go:
```diff
@@ -0,0 +1,240 @@
+//
+// The Reader is read until io.EOF.
+// Validation is performed to ensure that the DigestType is known.
+func NewDigestForContent(digestType DigestType, reader io.Reader) (Digest, error) {
```
Yep, done!

bufdev commented on private/bufpkg/bufcas/digest.go:
```diff
@@ -0,0 +1,240 @@
+//
+// Validation is performed to ensure the DigestType is known, and the value
+// is a valid digest value for the given DigestType.
+// TODO: validate the input proto Digest.
```
`protovalidate-go` validation, now added, and all Protobuf conversion functions now in `proto.go`.

bufdev commented on private/bufpkg/bufcas/file_node.go:
```diff
@@ -0,0 +1,140 @@
+
+// NewFileNode returns a new FileNode.
+//
+// The Digest may be nil.
```
Fixed.

bufdev commented on private/bufpkg/bufcas/file_set.go:
```diff
@@ -0,0 +1,191 @@
+//
+// The Manifest is guaranteed to exactly correlate with the Blobs in the BlobSet,
+// that is the Digests of the FileNodes in the Manifest will exactly match the
+// Digests in the Blobs. Note that some FileNodes may have empty Digests, in which
+// case there is no corresponding Blob (as the content is empty).
```
Fixed.

bufdev commented on private/bufpkg/bufcas/file_set.go:
```diff
@@ -0,0 +1,191 @@
+) error {
+	for _, fileNode := range fileSet.Manifest().FileNodes() {
+		writeObjectCloser, err := bucket.Put(ctx, fileNode.Path(), storage.PutWithAtomic())
+		if err != nil {
+			return err
+		}
+		blob := fileSet.BlobSet().GetBlob(fileNode.Digest())
+		if _, err := writeObjectCloser.Write(blob.Content()); err != nil {
+			return err
+		}
+		if err := writeObjectCloser.Close(); err != nil {
+			return err
+		}
```
Done

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+		if existingFileNode, ok := pathToFileNode[fileNode.Path()]; ok {
+			if !DigestEqual(existingFileNode.Digest(), fileNode.Digest()) {
+				return nil, fmt.Errorf("path %q had different Digests when constructing FileNode", fileNode.Path())
+			}
```
Changed.

bufdev commented on private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go:
```diff
@@ -46,41 +47,44 @@ func (c *casModuleCacher) GetModule(
+	blobs := make([]bufcas.Blob, 0, len(manifest.FileNodes()))
+	for _, fileNode := range manifest.FileNodes() {
+		fileDigest := fileNode.Digest()
+		if fileDigest == nil {
+			continue
+		}
```
Legacy from when `Digests` could be `nil` - removed.

bufdev commented on private/bufpkg/bufcas/file_node.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func (f *fileNode) String() string {
+	if f.digest == nil {
```
Every other entity in this package has a `String()` function that gives its canonical representation, there isn't a compelling reason to deviate specifically for some of the types. The rest of the codebase uses this pattern as well, as well as `Parse.*` functions. We don't have any `MarshalText` functions that I can see. I renamed `New.*ForString` to `Parse.*`.

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+
+// NewManifest returns a new Manifest for the given path -> Digest map.
+//
+// FileNodes are deduplicated upon construction, however if two FileNodes
```
It deduplicates the same as `BlobSet` does - it would seem OK to construct a `[]FileNode` via say walking a few buckets, and if there's a duplicate, filtering it out here. `BlobSet` we definitely want the deduplication`, here is more arguable, but we probably want to be consistent. Just my two cents.

bufdev replied:
I changed `NewBlobSet` to return an `error` to give us more flexibility in the future to add validation without needing to refactor.

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+// Maniest is a set FileNodes.
+type Manifest interface {
+	// fmt.Stringer encodes the Manifest into its canonical form, consisting of
+	// an ordered list of paths and their digests. Sorted by path.
```
It says that they are ordered, but does not specify what the ordering is. This comment makes it explicit (for example, it could be by digest).

bufdev commented on private/bufpkg/bufcas/digest.go:
```diff
@@ -0,0 +1,240 @@
+// TODO: validate the returned Digest.
+func DigestToProto(digest Digest) (*storagev1beta1.Digest, error) {
+	protoDigestType, ok := digestTypeToProto[digest.Type()]
+	// Technically we have aleady done this validation but just to be safe.
```
Done

bufdev commented on private/bufpkg/bufcas/file_set.go:
```diff
@@ -0,0 +1,191 @@
+//
+// Validation is done to ensure the Manifest exactly matches the BlobSet.
+// TODO: validate the input proto Blobs.
+func ProtoManifestBlobsAndBlobsToFileSet(
```
Fixed.

bufdev commented on private/bufpkg/bufcas/file_set.go:
```diff
@@ -0,0 +1,191 @@
+	return protoManifestBlob, protoBlobs, nil
+}
+
+// ManifestBlobsAndBlobsToFileSet converts the given manifest Blob and set of Blobs representing
```
Fixed.

bufdev commented on private/bufpkg/bufcas/blob.go:
```diff
@@ -0,0 +1,155 @@
+// We should be able to remove this, and this is commented out for now. Technically we do not
+// need to compare the contents, as we know that the Digest is a valid Digest for the
+// given content via valiation we did at construction time.
+func BlobEqual(a Blob, b Blob) bool {
```
An interface in Golang should contain exactly the minimal surface area it needs to expose the logic that is required to build on it, and nothing more (more easily implemented by others, easier if you need to do new interfaces that embed this interface type, etc). In this case, you can express equality based on the other methods on the interface, so it should be a separate function, not part of the interface itself.

Keeping this function for now, however, is a borderline case. I think it's worth having the package say what is considered an equal blob (and specifically, encoding that you can skip the content check), but I'm not sure if we need this right now. Will remove to minimize package surface area.

bufdev commented on private/bufpkg/bufcas/digest.go:
```diff
@@ -0,0 +1,240 @@
+	switch digestType {
+	case DigestTypeShake256:
+		shakeHash := sha3.NewShake256()
+		shakeHash.Reset()
```
Had just copied this from the previous implementation. Just out of extreme caution, can we keep this for now, and remove it later if we deem that this has no effect? I just looked at the underlying code, and while yes, according to the API, this `Reset` should have no effect, there's some stuff going on under the hood that concerns me. We may have some fire fights after this PR is merged, I'd like to eliminate this as a possible issue for us (by keeping it with a TODO to remove for now).

bufdev commented on private/bufpkg/bufcas/file_set.go:
```diff
@@ -0,0 +1,191 @@
+		}
+		blob := fileSet.BlobSet().GetBlob(fileNode.Digest())
+		if _, err := writeObjectCloser.Write(blob.Content()); err != nil {
+			return err
```
Good call - added.

bufdev commented on private/bufpkg/bufcas/file_set.go:
```diff
@@ -0,0 +1,191 @@
+}
+
+func newFileSet(manifest Manifest, blobSet BlobSet) *fileSet {
+	return &fileSet{
```
We generally don't do nil checking for parameters, see other comments. I think I added it for `FileNode` because I was already doing it for path, and path has some weird behavior on empty for `NormalizeAndValidate`. Not a great argument but.

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+	//	shake256:3b353aa5aacd11015e8577f16e2c4e7a242ce773d8e3a16806795bb94f76e601b0db9bf42d5e1907fda63303e1fa1c65f1c175ecc025a3ef29c3456ad237ad84  buf.md
+	//	shake256:7c88a20cf931702d042a4ddee3fde5de84814544411f1c62dbf435b1b81a12a8866a070baabcf8b5a0d31675af361ccb2d93ddada4cdcc11bab7ea3d8d7c4667  buf.yaml
+	//	shake256:9db25155eafd19b36882cff129daac575baa67ee44d1cb1fd3894342b28c72b83eb21aa595b806e9cb5344759bc8308200c5af98e4329aa83014dde99afa903a  pet/v1/pet.proto
+	fmt.Stringer
```
See response

bufdev commented on private/bufpkg/bufcas/manifest.go:
```diff
@@ -0,0 +1,183 @@
+	storagev1beta1 "github.com/bufbuild/buf/private/gen/proto/go/buf/registry/storage/v1beta1"
+)
+
+// Maniest is a set FileNodes.
```
Fixed.

bufdev commented on private/bufpkg/bufmodule/bufmodule.go:
```diff
@@ -133,21 +133,14 @@ type Module interface {
+	//
+	// FileSet will be nil if the Module was not constructed with a FileSet.
+	//
+	// A FileSets's contents contain a lexicographically sorted list of path names along
```
Fixed.

bufdev commented on private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go:
```diff
@@ -249,17 +228,3 @@ func (c *casModuleCacher) atomicWrite(ctx context.Context, contents io.Reader, p
-		return nil, err
-	}
-	defer func() {
-		retErr = multierr.Append(retErr, f.Close())
```
Done.

bufdev commented:
This should be ready to go now @saquibmian @pkwarren @unmultimedio 

- I addressed all comments.
- Tests were ported from the old `manifest` package.
- All tests are passing.
- I disabled `protovalidate` for now, we likely do not want to do this client-side, but there is an option to pass a validator.

Remaining:

- Some local QA.
- Porting of BSR.

pkwarren reviewed (approved):
Verified cache locally - everything working as before. Looks good.
